risc0/zkvm/src/mini_monitor.rs

```diff
@@ -0,0 +1,199 @@
+};
+use risc0_zkvm_platform::{
+    memory::{SYSTEM, TEXT_START},
+    syscall::{reg_abi::REG_MAX, DIGEST_BYTES},
+    WORD_SIZE,
+};
+use rrs_lib::Memory;
+use serde::{Deserialize, Serialize};
+
+use crate::sha::{Impl, Sha256};
+
+// each element of the merkle path has the following:
+// 1) a page entry address index. None indicates root index.
+// 2) a reference count
+// 3) the page bytes
+pub type MerklePathElement = (Option<u32>, u32, Vec<u8>);
+
+/// This is a small implementation of a memory monitor used for the fault
+/// checker. It records all memory regions needed for a signle RISC-V
+/// instruction execution.
+#[derive(Serialize, Deserialize, Debug)]
+pub struct MiniMonitor {
+    /// The program counter of the instruction to execute
+    pub pc: u32,
+    /// A mapping between all register values and program counter and a word of
+    /// data for values that are valid addresses.
+    pub memory_map: HashMap<u32, MerklePathElement>,
+    /// size of each page within each entry of the [MerklePathElement]
+    pub page_size: u32,
+}
+
+impl MiniMonitor {
+    const LOWEST: u32 = TEXT_START;
+    const HIGHEST: u32 = SYSTEM.start() as u32;
+    /// Given an address, return true if the address is within the zkVM's
+    /// accessible memory regions
+    pub fn address_is_within_range(&self, addr: u32) -> bool {
+        Self::LOWEST <= addr && addr < Self::HIGHEST
+    }
+
+    /// get the 32-bit value value referenced by the program counter.
+    pub fn get_pc_value(&self) -> Result<u32> {
+        let offset = self.get_offset(self.pc) as usize;
+        let idx = self.get_page_index(self.pc);
+
+        let page = match self.memory_map.get(&idx) {
+            Some((_, _, data)) => data,
+            None => bail!("page index {idx} does not exist in the memory map"),
+        };
+
+        if offset + WORD_SIZE > page.len() {
+            bail!("Not enough data in page");
+        }
+        Ok(*bytemuck::from_bytes::<u32>(
+            &page[offset..offset + WORD_SIZE],
+        ))
+    }
+
+    /// get all register contents from the memory map.
+    pub fn get_registers(&self) -> Result<[u32; REG_MAX]> {
+        const REGISTER_LENGTH: usize = WORD_SIZE * REG_MAX;
+        let offset = self.get_offset(SYSTEM.start() as u32) as usize;
+        let idx = self.get_page_index(SYSTEM.start() as u32);
+
+        let page = match self.memory_map.get(&idx) {
+            Some((_, _, data)) => data,
+            None => bail!("page index {idx} does not exist in the memory map"),
+        };
+
+        if offset + REGISTER_LENGTH > page.len() {
+            bail!("Not enough data in SYSTEM region");
+        }
+
+        let mm_regs: &[u32] = bytemuck::cast_slice(&page[offset..offset + REGISTER_LENGTH]);
+        let mut registers = [0u32; REG_MAX];
+
+        for idx in 0..REG_MAX {
+            registers[idx] = mm_regs[idx];
+        }
+        Ok(registers)
+    }
+
+    fn hash_page_bytes(page: &[u8]) -> Digest {
```

Comment:
There's a lot of logic that I borrowed from monitor.rs and image.rs here. Including this function which is a duplicate in the image.rs file. There were some tricky elements to this because the sha Impl resides in the zkvm crate but we can't have risc0-binfmt and risc0-build depend on the risc0-zkvm. I'm thinking that we may want to refactor this part...